diff --git a/tools/godotcpp.py b/tools/godotcpp.py
index 60cd34b..676b626 100644
--- a/tools/godotcpp.py
+++ b/tools/godotcpp.py
@@ -284,8 +284,8 @@ def _godot_cpp(env):
     )
     # Forces bindings regeneration.
     if env["generate_bindings"]:
-        AlwaysBuild(bindings)
-        NoCache(bindings)
+        env.AlwaysBuild(bindings)
+        env.NoCache(bindings)
 
     # Sources to compile
     sources = []
