diff --git a/tools/web.py b/tools/web.py
index c8f07c5..010efac 100644
--- a/tools/web.py
+++ b/tools/web.py
@@ -45,6 +45,9 @@ def generate(env):
     # Force wasm longjmp mode.
     env.Append(CCFLAGS=["-sSUPPORT_LONGJMP='wasm'"])
     env.Append(LINKFLAGS=["-sSUPPORT_LONGJMP='wasm'"])
+
+    # Use big int
+    env.Append(LINKFLAGS=["-sWASM_BIGINT"])
 
     env.Append(CPPDEFINES=["WEB_ENABLED", "UNIX_ENABLED"])
 
