diff --git a/tools/web.py b/tools/web.py
index c491880..fba7eb7 100644
--- a/tools/web.py
+++ b/tools/web.py
@@ -46,6 +46,9 @@ def generate(env):
     env.Append(CCFLAGS=["-sSUPPORT_LONGJMP='wasm'"])
     env.Append(LINKFLAGS=["-sSUPPORT_LONGJMP='wasm'"])
 
+    # Use big int
+    env.Append(LINKFLAGS=["-sWASM_BIGINT"])
+
     env.Append(CPPDEFINES=["WEB_ENABLED", "UNIX_ENABLED"])
 
     # Refer to https://github.com/godotengine/godot/blob/master/platform/web/detect.py
