diff --git a/tools/javascript.py b/tools/javascript.py
index 42c601d..8fcb921 100644
--- a/tools/javascript.py
+++ b/tools/javascript.py
@@ -1,18 +1,14 @@
 import os
-
+from SCons.Util import WhereIs
 
 def exists(env):
-    return "EM_CONFIG" in os.environ
-
+    return WhereIs("emcc") is not None
 
 def generate(env):
     if env["arch"] not in ("wasm32"):
         print("Only wasm32 supported on web. Exiting.")
         Exit()
 
-    if "EM_CONFIG" in os.environ:
-        env["ENV"] = os.environ
-
     env["CC"] = "emcc"
     env["CXX"] = "em++"
     env["AR"] = "emar"
@@ -26,6 +22,13 @@ def generate(env):
     env["ARCOM_POSIX"] = env["ARCOM"].replace("$TARGET", "$TARGET.posix").replace("$SOURCES", "$SOURCES.posix")
     env["ARCOM"] = "${TEMPFILE(ARCOM_POSIX)}"
 
+    # Thread support (via SharedArrayBuffer).
+    env.Append(CPPDEFINES=["PTHREAD_NO_RENAME"])
+    env.Append(CCFLAGS=["-s", "USE_PTHREADS=1"])
+    env.Append(LINKFLAGS=["-s", "USE_PTHREADS=1"])
+    env.Append(LINKFLAGS=["-s", "PTHREAD_POOL_SIZE=8"])
+    env.Append(LINKFLAGS=["-s", "WASM_MEM_MAX=2048MB"])
+
     # All intermediate files are just LLVM bitcode.
     env["OBJPREFIX"] = ""
     env["OBJSUFFIX"] = ".bc"
@@ -39,9 +42,9 @@ def generate(env):
     env.Replace(SHLINKFLAGS="$LINKFLAGS")
     env.Replace(SHLINKFLAGS="$LINKFLAGS")
 
-    if env["target"] == "debug":
+    if env["target"] in ["editor", "template_debug"]:
         env.Append(CCFLAGS=["-O0", "-g"])
-    elif env["target"] == "release":
+    elif env["target"] == "template_release":
         env.Append(CCFLAGS=["-O3"])
 
     env.Append(CPPDEFINES=["WEB_ENABLED", "UNIX_ENABLED"])
