diff --git a/binding_generator.py b/binding_generator.py
index 49664d1..eb940a6 100644
--- a/binding_generator.py
+++ b/binding_generator.py
@@ -1,5 +1,9 @@
 #!/usr/bin/env python
 
+import sys
+sys.path.insert(0, "../patches")
+import get_used_classes
+
 import json
 import re
 import shutil
@@ -76,6 +80,7 @@ def get_file_list(api_filepath, output_dir, headers=False, sources=False):
     with open(api_filepath, encoding="utf-8") as api_file:
         api = json.load(api_file)
 
+    get_used_classes.scan_dependencies(api)
     core_gen_folder = Path(output_dir) / "gen" / "include" / "godot_cpp" / "core"
     include_gen_folder = Path(output_dir) / "gen" / "include" / "godot_cpp"
     source_gen_folder = Path(output_dir) / "gen" / "src"
@@ -132,6 +137,7 @@ def get_file_list(api_filepath, output_dir, headers=False, sources=False):
         register_engine_classes_source_path = source_gen_folder / "register_engine_classes.cpp"
         files.append(str(register_engine_classes_source_path.as_posix()))
 
+    files = get_used_classes.delete_useless(files)
     return files
 
 
@@ -142,6 +148,8 @@ def print_file_list(api_filepath, output_dir, headers=False, sources=False):
 
 
 def scons_emit_files(target, source, env):
+    get_used_classes.setup(env.get("exclude_unused_classes", True), env.get("folder_to_include_classes", "../src"))
+
     files = [env.File(f) for f in get_file_list(str(source[0]), target[0].abspath, True, True)]
     env.Clean(target, files)
     env["godot_cpp_gen_dir"] = target[0].abspath
@@ -1599,6 +1607,9 @@ def generate_register_engine_classes_source(api):
         if class_api["name"] == "ClassDB":
             continue
 
+        if get_used_classes.need_to_exclude(class_api["name"]):
+            continue
+
         class_name = class_api["name"]
         snake_class_name = camel_to_snake(class_name)
 
diff --git a/tools/godotcpp.py b/tools/godotcpp.py
index 60cd34b..8013462 100644
--- a/tools/godotcpp.py
+++ b/tools/godotcpp.py
@@ -174,6 +174,9 @@ def options(opts, env):
         )
     )
 
+    opts.Add(BoolVariable("exclude_unused_classes", "Disable generation of unused classes.", True))
+    opts.Add(PathVariable("folder_to_include_classes", "Path to the directory containing extension sources", "../src", PathVariable.PathIsDir))
+
     # Add platform options
     for pl in platforms:
         tool = Tool(pl, toolpath=["tools"])
