diff --git a/tools/godotcpp.py b/tools/godotcpp.py
index 60cd34b..6815eaa 100644
--- a/tools/godotcpp.py
+++ b/tools/godotcpp.py
@@ -6,6 +6,8 @@ from SCons.Builder import Builder
 
 from binding_generator import scons_generate_bindings, scons_emit_files
 
+sys.path.insert(0, "../patches")
+import unity_tools
 
 def add_sources(sources, dir, extension):
     for f in os.listdir(dir):
@@ -282,6 +284,9 @@ def _godot_cpp(env):
             "binding_generator.py",
         ],
     )
+
+    bindings = unity_tools.generate_unity_build(bindings, "godot-cpp_")
+
     # Forces bindings regeneration.
     if env["generate_bindings"]:
         env.AlwaysBuild(bindings)
