diff --git a/tools/godotcpp.py b/tools/godotcpp.py
index cc2b02f..dbeecce 100644
--- a/tools/godotcpp.py
+++ b/tools/godotcpp.py
@@ -267,6 +267,8 @@ def options(opts, env):
         )
     )
 
+    opts.Add(BoolVariable(key="threads", help="Enable threading support", default=env.get("threads", True)))
+
     # compiledb
     opts.Add(
         BoolVariable(
@@ -429,6 +431,8 @@ def generate(env):
 
     # Suffix
     suffix = ".{}.{}".format(env["platform"], env["target"])
+    if not env["threads"]:
+        suffix += ".nothreads"
     if env.dev_build:
         suffix += ".dev"
     if env["precision"] == "double":
diff --git a/tools/web.py b/tools/web.py
index 79b4b24..8953325 100644
--- a/tools/web.py
+++ b/tools/web.py
@@ -34,8 +34,9 @@ def generate(env):
     env["SHLIBSUFFIX"] = ".wasm"
 
     # Thread support (via SharedArrayBuffer).
-    env.Append(CCFLAGS=["-s", "USE_PTHREADS=1"])
-    env.Append(LINKFLAGS=["-s", "USE_PTHREADS=1"])
+    if env["threads"]:
+        env.Append(CCFLAGS=["-s", "USE_PTHREADS=1"])
+        env.Append(LINKFLAGS=["-s", "USE_PTHREADS=1"])
 
     # Build as side module (shared library).
     env.Append(CPPFLAGS=["-s", "SIDE_MODULE=1"])
